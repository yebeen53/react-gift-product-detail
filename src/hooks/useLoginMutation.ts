import { useMutation } from '@tanstack/react-query';
import { apiClient } from '@/api/apiClient';

interface LoginPayload {
  email: string;
  password: string;
}

interface LoginResponse {
  authToken: string;
  id: number;
  name: string;
  email: string;
}

export const useLoginMutation = () => {
  return useMutation({
    mutationFn: async (payload: LoginPayload): Promise<LoginResponse> => {
      const res = await apiClient.post('/api/login', payload);
      return res.data.data;
    },
  });
};