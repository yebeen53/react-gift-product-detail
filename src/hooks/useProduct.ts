import {
  useSuspenseQuery,
  useMutation,
  useQueryClient,
} from '@tanstack/react-query';
import {
  fetchProduct,
  fetchProductDetail,
  fetchHighlightReview,
  fetchWishCount,
} from '@/api/products/product';

export const useProduct = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['product', productId],
    queryFn: () => fetchProduct(productId),
  });

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });

export const useHighlightReview = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['highlightReview', productId],
    queryFn: () => fetchHighlightReview(productId),
  });

export const useWishCount = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['wishCount', productId],
    queryFn: () => fetchWishCount(productId),
  });

export const useWishMutation = (productId: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async () => Promise.resolve(), // 실제 API 요청 함수 넣기

    onMutate: async () => {
      await queryClient.cancelQueries({ queryKey: ['wishCount', productId] });

      const prevWish = queryClient.getQueryData<{
        wishCount: number;
        isWished: boolean;
      }>(['wishCount', productId]);

      if (prevWish) {
        queryClient.setQueryData(['wishCount', productId], {
          ...prevWish,

          wishCount: prevWish.isWished
            ? prevWish.wishCount - 1
            : prevWish.wishCount + 1,
          isWished: !prevWish.isWished, // 토글
        });
      }

      return { prevWish };
    },

    onError: (_err, _vars, context) => {
      if (context?.prevWish) {
        queryClient.setQueryData(['wishCount', productId], context.prevWish);
      }
    },

    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['wishCount', productId] });
    },
  });
};
