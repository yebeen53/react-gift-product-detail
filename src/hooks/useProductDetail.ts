import { useSuspenseQuery } from '@tanstack/react-query';
import axios from 'axios';

interface Announcement {
  name: string;
  value: string;
  displayOrder: number;
}

interface ProductDetail {
  description: string;
  announcements: Announcement[];
}

const fetchProductDetail = async (productId: string): Promise<ProductDetail> => {
  const response = await axios.get(`/api/products/${productId}/detail`);
  return response.data.data;
};

export const useProductDetail = (productId: string) =>
  useSuspenseQuery({
    queryKey: ['productDetail', productId],
    queryFn: () => fetchProductDetail(productId),
  });
