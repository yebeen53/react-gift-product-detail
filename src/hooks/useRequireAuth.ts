import useAuth from '@/hooks/useAuth';
import { useLocation, useNavigate } from 'react-router-dom';
import { useEffect } from 'react';
import { ROUTES } from '@/constants/routes';

const useRequireAuth = () => {
  const { user } = useAuth();
  const location = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    if (!user || !user.authToken) {
      navigate(ROUTES.HOME, {
        replace: true,
        state: { from: location.pathname },
      });
    }
  }, [user, navigate, location]);

  return user;
};

export default useRequireAuth;
