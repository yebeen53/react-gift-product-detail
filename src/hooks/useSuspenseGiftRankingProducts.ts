import { fetchRanking } from '@/api/products/ranking';
import { useSuspenseQuery } from '@tanstack/react-query';

export const useSuspenseGiftRankingProducts = (
  genderCode: string,
  categoryCode: string
) => {
  return useSuspenseQuery({
    queryKey: ['giftRankingProducts', genderCode, categoryCode],
    queryFn: () => fetchRanking(genderCode, categoryCode),
  });
};
