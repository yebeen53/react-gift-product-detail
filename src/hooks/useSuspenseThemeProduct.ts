import {
  useSuspenseInfiniteQuery,
  type QueryFunctionContext,
} from '@tanstack/react-query';
import axios from 'axios';
import type { ThemeProductsData } from '@/types/product';

const fetchThemeProducts = async (
  ctx: QueryFunctionContext<readonly [string, string], number>
): Promise<ThemeProductsData> => {
  const [, themeId] = ctx.queryKey;
  const pageParam = ctx.pageParam ?? 0;

  const res = await axios.get(`/api/themes/${themeId}/products`, {
    params: { cursor: pageParam, limit: 10 },
  });
  const data = res.data.data;

  return {
    list: data.list,
    cursor: data.cursor,
    hasMoreList: data.hasMoreList,
  };
};

export const useSuspenseThemeProducts = (themeId: string) => {
  return useSuspenseInfiniteQuery({
    queryKey: ['themeProducts', themeId] as const,
    queryFn: fetchThemeProducts,
    initialPageParam: 0,
    getNextPageParam: (lastPage) =>
      lastPage.hasMoreList ? lastPage.cursor : undefined,
  });
};
