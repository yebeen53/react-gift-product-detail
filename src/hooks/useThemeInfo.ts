import { apiClient } from '@/api/apiClient';
import { useSuspenseQuery } from '@tanstack/react-query';

export interface ThemeDetail {
  image: string | undefined;
  name: string;
  themeId: string;
  title: string;
  description: string | null;
  backgroundColor: string;
}

const fetchThemeDetail = async (themeId: string): Promise<ThemeDetail> => {
  const res = await apiClient.get(`/api/themes/${themeId}/info`);
  const data = res.data?.data;
  if (!data) {
    throw new Error('테마 정보가 없습니다.');
  }
  return data;
};

export const useThemeInfo = (themeId: string) => {
  return useSuspenseQuery({
    queryKey: ['themeDetail', themeId],
    queryFn: () => fetchThemeDetail(themeId),
  });
};
