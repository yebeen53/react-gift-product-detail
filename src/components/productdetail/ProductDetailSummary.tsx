import theme from '@/data/theme';
import { useProduct } from '@/hooks/useProduct';

interface Props {
  productId: string;
}

const ProductDetailSummary = ({ productId }: Props) => {
  const { data: product } = useProduct(productId);

  return (
    <section>
      <img src={product.imageURL} />
      <h1 style={{ margin: theme.spacing.spacing4,}}>{product.name}</h1>
      <h1 style={{ margin: theme.spacing.spacing4, }}>{product.price.sellingPrice}Ïõê</h1>
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          gap: theme.spacing.spacing2,
          margin: theme.spacing.spacing4,
        }}
      >
        <img
          src={product.brandInfo.imageURL}
          alt={product.brandInfo.name}
          style={{ height: theme.spacing.spacing8, width: 'auto', objectFit: 'contain' }}
        />
        <h1 style={{ fontSize: theme.spacing.spacing8, margin: 0 }}>
          {product.brandInfo.name}
        </h1>
      </div>
    </section>
  );
};

export default ProductDetailSummary;
