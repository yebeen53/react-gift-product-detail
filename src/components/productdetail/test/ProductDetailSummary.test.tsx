import { render, screen, waitFor } from '@testing-library/react';
import ProductDetailSummary from '../ProductDetailSummary';
import { useProduct } from '@/hooks/useProduct';

jest.mock('@/hooks/useProduct', () => ({
  useProduct: jest.fn(),
}));

describe('ProductDetailSummary 컴포넌트', () => {
  const mockProductData = {
    imageURL: 'product-main.jpg',
    name: '최신 스마트워치',
    price: { sellingPrice: 250000 },
    brandInfo: {
      imageURL: 'brand-logo.png',
      name: 'ABC 전자',
    },
  };

  beforeEach(() => {
    (useProduct as jest.Mock).mockClear();
  });

  it('제품 정보가 올바르게 렌더링되어야 한다', async () => {
    (useProduct as jest.Mock).mockReturnValue({ data: mockProductData });

    render(<ProductDetailSummary productId="p1" />);

    await waitFor(() => {
      const productImage = screen.getByRole('img', { name: mockProductData.name });
      expect(productImage).toHaveAttribute('src', mockProductData.imageURL);

      const productName = screen.getByRole('heading', { name: mockProductData.name });
      expect(productName).toBeInTheDocument();

      const priceText = `${mockProductData.price.sellingPrice}원`;
      const priceHeading = screen.getByRole('heading', { name: priceText });
      expect(priceHeading).toBeInTheDocument();

      const brandLogo = screen.getByRole('img', { name: mockProductData.brandInfo.name });
      expect(brandLogo).toHaveAttribute('src', mockProductData.brandInfo.imageURL);
      const brandName = screen.getByRole('heading', { name: mockProductData.brandInfo.name });
      expect(brandName).toBeInTheDocument();
    });
  });

  it('컴포넌트의 스타일 및 구조가 스냅샷과 일치해야 한다', async () => {
    (useProduct as jest.Mock).mockReturnValue({ data: mockProductData });

    const { asFragment } = render(<ProductDetailSummary productId="p1" />);
    await waitFor(() => {}); 

    expect(asFragment()).toMatchSnapshot();
  });
});
