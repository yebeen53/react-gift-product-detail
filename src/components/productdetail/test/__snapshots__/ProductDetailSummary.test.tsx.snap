// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductDetailSummary 컴포넌트 컴포넌트의 스타일 및 구조가 스냅샷과 일치해야 한다 1`] = `
<DocumentFragment>
  <section>
    <img
      alt="최신 스마트워치"
      src="product-main.jpg"
    />
    <h1
      style="margin: 16px;"
    >
      최신 스마트워치
    </h1>
    <h1
      style="margin: 16px;"
    >
      250000원
    </h1>
    <div
      style="display: flex; align-items: center; gap: 8px; margin: 16px;"
    >
      <img
        alt="ABC 전자"
        src="brand-logo.png"
        style="height: 32px; width: auto; object-fit: contain;"
      />
      <h1
        style="font-size: 32px; margin: 0px;"
      >
        ABC 전자
      </h1>
    </div>
  </section>
</DocumentFragment>
`;
